File: scala/dllib/src/main/scala/com/intel/analytics/bigdl/dllib/example/loadmodel/ModelValidator.scala
/*
 * Copyright 2016 The BigDL Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.intel.analytics.bigdl.example.loadmodel

import java.nio.file.Paths

import com.intel.analytics.bigdl.models.inception.Inception_v1_NoAuxClassifier
import com.intel.analytics.bigdl.nn.Module
import com.intel.analytics.bigdl.optim.{Top1Accuracy, Top5Accuracy, Validator}
import com.intel.analytics.bigdl.utils.Engine
import org.apache.log4j.Logger
import org.apache.spark.SparkContext
import scopt.OptionParser

import scala.language.existentials

/**
 * ModelValidator provides an integrated example to load models,
 * and test over imagenet validation dataset
 * (running as a local Java program, or a standard Spark program).
 */
object ModelValidator {

  val logger = Logger.getLogger(getClass)

  /**
   * This is a trait meaning the model type.
   * There are three sorts of model type, which
   * are torch model [[TorchModel]], caffe model
   * [[CaffeModel]] and BigDL model [[BigDlModel]].
   */
  sealed trait ModelType

  case object TorchModel extends ModelType

  case object CaffeModel extends ModelType

  case object BigDlModel extends ModelType

  case class TestLocalParams(
    folder: String = "./",
    modelType: ModelType = null,
    modelName: String = "",
    caffeDefPath: Option[String] = None,
    modelPath: String = "",
    batchSize: Int = 32,
    meanFile: Option[String] = None
  )

  val testLocalParser = new OptionParser[TestLocalParams]("BigDL Load Model Example") {
    head("BigDL Load Model Example")
    opt[String]('f', "folder")
      .text("where you put your local image files")
      .action((x, c) => c.copy(folder = x))
    opt[String]('m', "modelName")
      .text("the model name you want to test")
      .required()
      .action((x, c) => c.copy(modelName = x.toLowerCase()))
    opt[String]('t', "modelType")
      .text("torch, caffe or bigdl")
      .required()
      .action((x, c) =>
        x.toLowerCase() match {
          case "torch" => c.copy(modelType = TorchModel)
          case "caffe" => c.copy(modelType = CaffeModel)
          case "bigdl" => c.copy(modelType = BigDlModel)
          case _ =>
            throw new IllegalArgumentException("only torch, caffe or bigdl supported")
        }
      )
    opt[String]("caffeDefPath")
      .text("caffe define path")
      .action((x, c) => c.copy(caffeDefPath = Some(x)))
    opt[String]("modelPath")
      .text("model path")
      .action((x, c) => c.copy(modelPath = x))
    opt[Int]('b', "batchSize")
      .text("batch size")
      .action((x, c) => c.copy(batchSize = x))
    opt[String]("meanFile")
      .text("mean file")
      .action((x, c) => c.copy(meanFile = Some(x)))
  }

  def main(args: Array[String]): Unit = {
    testLocalParser.parse(args, TestLocalParams()).foreach(param => {
      val conf = Engine.createSparkConf()
      conf.setAppName("BigDL Image Classifier Example")
      val sc = new SparkContext(conf)
      Engine.init

      val valPath = param.folder

      val (model, validateDataSet) = param.modelType match {
        case CaffeModel =>
          param.modelName match {
            case "alexnet" =>
              (Module.loadCaffeModel[Float](
                param.caffeDefPath.get, param.modelPath),
                AlexNetPreprocessor.rdd(valPath, param.batchSize, param.meanFile.get, sc))
            case "inception" =>
              (Module.loadCaffeModel[Float](
                param.caffeDefPath.get, param.modelPath),
                InceptionPreprocessor.rdd(valPath, param.batchSize, sc))
          }

        case TorchModel =>
          param.modelName match {
            case "resnet" =>
              (Module.loadTorch[Float](param.modelPath),
                ResNetPreprocessor.rdd(valPath, param.batchSize, sc))
          }

        case BigDlModel =>
          param.modelName match {
            case "resnet" =>
              (Module.loadModule[Float](param.modelPath),
                ResNetPreprocessor.rdd(valPath, param.batchSize, sc, BigDlModel))
            case "vgg16" =>
              (Module.loadModule[Float](param.modelPath),
                VGGPreprocessor.rdd(valPath, param.batchSize, sc))
          }

        case _ => throw new IllegalArgumentException(s"${ param.modelType } is not" +
          s"supported in this example, please use alexnet/inception/resnet")
      }
      println(model)

      val result = model.evaluate(
        validateDataSet,
        Array(new Top1Accuracy[Float](), new Top5Accuracy[Float]()),
        Some(param.batchSize))

      result.foreach(r => {
        logger.info(s"${ r._2 } is ${ r._1 }")
      })
      sc.stop()
    })
  }
}

File: pyspark/dl/models/loadmodel/modelvalidator.py
#
# Copyright 2016 The BigDL Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# Still in experimental stage!

from optparse import OptionParser

from dataset.transformer import *
from dataset import imagenet
from models.inception import inception
from models.loadmodel.alexnet import alexnet
from nn.layer import *
from util.common import *

def get_data(folder, file_type="local", normalize=255.0):
    if "hdfs" == file_type:
        return imagenet.read_seq_file(sc, folder, normalize)
    elif "local" == file_type:
        return imagenet.read_local(sc, folder, 255.0)

if __name__ == "__main__":
    parser = OptionParser()
    parser.add_option("-f", "--folder", dest="folder", default="./")
    parser.add_option("-m", "--modelName", dest="modelName", default="")
    parser.add_option("-t", "--modelType", dest="modelType", default="")
    parser.add_option("--caffeDefPath", dest="caffeDefPath", default="")
    parser.add_option("--modelPath", dest="modelPath", default="")
    parser.add_option("-b", "--batchSize", dest="batchSize", default="32")
    parser.add_option("--meanFile", dest="meanFile", default="imagenet_mean.binaryproto")

    (options, args) = parser.parse_args(sys.argv)

    sc = SparkContext(appName="image_classifier",
                      conf=create_spark_conf())
    init_engine()

    if options.modelType == "caffe":
        if options.modelName == "alexnet":
            image_size = 227
            model = Model.load_caffe(alexnet.alexnet(1000), options.caffeDefPath, options.modelPath)
            means = imagenet.load_mean_file(options.meanFile)
            validate_dataset = get_data(options.folder, "hdfs").map(
                pixel_normalizer(means)).map(
                crop(image_size, image_size, "center")).map(
                transpose(False))

        elif options.modelName == "inception":
            image_size = 224
            model = Model.load_caffe(inception.Inception_v1_NoAuxClassifier(1000), options.caffeDefPath, options.modelPath)
            validate_dataset = get_data(options.folder, "hdfs", 1.0).map(
                crop(image_size, image_size, "center")).map(
                channel_normalizer(123, 117, 104, 1, 1, 1)).map(
                transpose(False))

    elif options.modelType == "torch":
        if options.modelName == "resnet":
            image_size = 224
            model = Model.load_torch(options.modelPath)
            validate_dataset = get_data(options.folder, "hdfs").map(
                crop(image_size, image_size, "center")).map(
                channel_normalizer(0.485, 0.456, 0.406, 0.229, 0.224, 0.225)).map(
                transpose(True))

    results = model.test(validate_dataset, int(options.batchSize), ["Top1Accuracy"])
    for result in results:
        print result
    sc.stop()
